be/src/exec/spill/spilled_stream.cpp

```diff
@@ -36,16 +36,263 @@
+        : _streams(std::move(streams)), _merger(state) {}
+
+Status SortedFileStream::init(const SortExecExprs* sort_exprs, const SortDescs* descs) {
+    std::vector<ChunkProvider> providers;
+    for (auto& stream : _streams) {
+        _buffered_streams.emplace_back(std::make_unique<BufferedSpilledStream>(chunk_buffer_max_size, stream));
+        auto buffered_stream = _buffered_streams.back();
+        auto chunk_provider = [buffered_stream, this](ChunkUniquePtr* out_chunk, bool* eos) {
+            if (out_chunk == nullptr || eos == nullptr) {
+                return buffered_stream->has_chunk();
+            }
+
+            if (!buffered_stream->has_chunk()) {
+                return false;
+            }
+
+            auto stv = buffered_stream->read_from_buffer();
+            // auto stv = buffered_stream->read(_read_context);
+            if (!stv.status().ok()) {
+                if (!stv.status().is_end_of_file()) {
+                    _status.update(stv.status());
+                }
+                buffered_stream->mark_is_eof();
+                *eos = true;
+                return false;
+            }
+
+            *out_chunk = std::move(stv.value());
+            return true;
+        };
+        providers.emplace_back(std::move(chunk_provider));
+    }
+    RETURN_IF_ERROR(_merger.init(providers, &(sort_exprs->lhs_ordering_expr_ctxs()), *descs));
+    return Status::OK();
+}
+
+StatusOr<ChunkUniquePtr> SortedFileStream::read(SpillFormatContext& context) {
+    ChunkUniquePtr chunk;
+    bool should_exit = false;
+    RETURN_IF_ERROR(_merger.get_next(&chunk, &_eos, &should_exit));
+
+    if (chunk && !chunk->is_empty()) {
+        return std::move(chunk);
+    }
+
+    if (_eos) {
+        return Status::EndOfFile("eos");
+    }
+
+    // data is not arrived
+    DCHECK(should_exit);
+    return std::make_unique<Chunk>();
+}
+
+class BufferedSpillReadTask final : public SpillRestoreTask {
+public:
+    BufferedSpillReadTask(std::weak_ptr<SpillerFactory> factory,
+                          std::vector<std::shared_ptr<BufferedSpilledStream>> streams)
+            : _factory(std::move(factory)), _buffered_stream(std::move(streams)) {}
+
+    BufferedSpillReadTask(std::weak_ptr<SpillerFactory> factory, std::shared_ptr<BufferedSpilledStream> buffered_stream)
+            : _factory(std::move(factory)), _buffered_stream({std::move(buffered_stream)}) {}
+
+    Status do_read(SpillFormatContext& context) override;
+
+private:
+    std::weak_ptr<SpillerFactory> _factory;
+    std::vector<std::shared_ptr<BufferedSpilledStream>> _buffered_stream;
+};
+
+Status BufferedSpillReadTask::do_read(SpillFormatContext& context) {
+    size_t eofs = 0;
+    for (auto& stream : _buffered_stream) {
+        if (!stream->buffer_fulled()) {
+            if (stream->acquire()) {
+                RETURN_IF_ERROR(stream->read_to_buffer(context));
+                stream->release();
+            } else {
+                DCHECK(false);
+            }
```

Comment:
1. what will happen if stream->acquire return false?
2. if `stream->read_to_buffer` returns error, `stream->release()` won't be invoked, is this as expected?

Reply:
1. now our stream is not thread safe. so stream->acquire() is always true. 
3. stream->read_to_buffer return error means we need abort the query.